capnpc/src/codegen.rs

```diff
@@ -240,8 +257,15 @@ impl<'a> GeneratorContext<'a> {
                     "{}_capnp",
                     path_to_stem_string(importpath)?.replace('-', "_")
                 );
+                let parent_module_scope = if let Some(krate) = crates_provide.get(&import.get_id())
```

Comment:
What should happen if a file in `ctx.request.get_requested_files()` has this ID too? Then the user is generating code for that file, and it seems to me that the crate-local version should win.

I think that, given an import ID `X`, the precedence should be:
1. If the user explicitly requests codegen a file with ID `X`, then that locally-generated version should be used, regardless of the contents of `crates_provide_map`.
2. Otherwise, if `X` appears in `crates_provide_map`, then that version of the name should be used.
3. Otherwise, if the referred-to file has a `$Rust.crate` annotation, then the `Text` crate name in that annotation should be used.
4. Otherwise, compile-time error (or graceful fallback).

I see that you removed the `$Rust.crate` annotation from this pull request. I do think we should add that annotation. Up to you whether you want it to be in this pull request or not.

Reply:
I'm tempted to say that if a user explicitly says a file should be provided by a crate and also explicitly asks to compile that file themselves we should fail with an error.

My guess is the main reason this would happen is that somebody was confused and thought they had to compile their imports. I can't think of a good reason you'd ever do this.

I agree $Rust.crate should be overridden by the build script option. I removed it because I was thinking it'd only be useful once we had the cargo feature but on second thought I think it's useful right away. I'll add it back.

Reply:
> I'm tempted to say that if a user explicitly says a file should be provided by a crate and also explicitly asks to compile that file themselves we should fail with an error.

Yeah that makes sense. I was forgetting that they need to explicitly opt into each individual file in `crate_provides()`.

Reply:
In terms of fallbacks I don't think it's that important. c++.capnp doesn't have any annotations with non built-in types, so I think it's unlikely users will have such annotations in their code and not want to parse the annotations in Rust.

If it's a field instead of an annotation I can't think of a use case where a fallback would be useful. 

Another possible API would be to fail with a helpful error message that suggests adding an explicit exclude_annotations(file_id) call.

Reply:
I was imagining a schema C++ project, where it has a bunch of $Json annotations sprinkled in. A Rust project might not care about JSON might to want to add a dependency on the `capnp_json` crate. So I think some way to ignore the annotations is needed. However, since there's probably more discussion to be had about it, and since nobody is actually complaining about it yet, I think it does not belong in the current PR.